File: SRM-KART/backend/src/main/java/com/srmkart/service/ListingService.java
package com.srmkart.service;

import com.srmkart.dao.ListingDAO;
import com.srmkart.model.Listing;
import java.util.List;

public class ListingService {
    private ListingDAO listingDAO = new ListingDAO();

    public List<Listing> getAllListings() {
        return listingDAO.getAllListings();
    }

    public Listing getListingById(int id) {
        return listingDAO.findById(id);
    }

    public boolean createListing(Listing listing) {
        return listingDAO.createListing(listing);
    }

    public List<Listing> searchListings(String query, String categoryId, Double minPrice, Double maxPrice) {
        return listingDAO.searchListings(query, categoryId, minPrice, maxPrice);
    }
}

File: SRM-KART/frontend/src/app/services/listing.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AuthService } from './auth.service';

export interface Listing {
  id?: number;
  title: string;
  description: string;
  price: number;
  categoryId: number;
  conditionStatus: string;
  sellerName?: string;
  categoryName?: string;
  imageUrl?: string;
  createdAt?: string;
}

export interface Category {
  id: number;
  name: string;
  icon: string;
}

@Injectable({
  providedIn: 'root'
})
export class ListingService {
  private apiUrl = 'http://localhost:8080/srmkart/api';

  constructor(private http: HttpClient, private authService: AuthService) {}

  private getAuthHeaders(): HttpHeaders {
    return new HttpHeaders({
      'Authorization': `Bearer ${this.authService.getToken()}`
    });
  }

  getAllListings(): Observable<Listing[]> {
    return this.http.get<Listing[]>(`${this.apiUrl}/listings`);
  }

  searchListings(query: string, categoryId: string = ''): Observable<Listing[]> {
    return this.http.get<Listing[]>(`${this.apiUrl}/search?q=${query}&category=${categoryId}`);
  }

  getCategories(): Observable<Category[]> {
    return this.http.get<Category[]>(`${this.apiUrl}/categories`);
  }

  getListingById(id: number): Observable<Listing> {
    return this.http.get<Listing>(`${this.apiUrl}/listings/${id}`);
  }

  createListing(listing: Listing): Observable<Listing> {
    return this.http.post<Listing>(`${this.apiUrl}/listings`, listing, { headers: this.getAuthHeaders() });
  }

  createListingWithImage(formData: FormData): Observable<Listing> {
    return this.http.post<Listing>(`${this.apiUrl}/listings`, formData, { headers: this.getAuthHeaders() });
  }
  
  getPlatformStats(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/stats`);
  }
}
